source/common/router/router.cc

```diff
@@ -1680,11 +1680,15 @@ void Filter::onUpstreamHeaders(uint64_t response_code, Http::ResponseHeaderMapPt
   if (DateUtil::timePointValid(downstream_request_complete_time_)) {
     Event::Dispatcher& dispatcher = callbacks_->dispatcher();
     MonotonicTime response_received_time = dispatcher.timeSource().monotonicTime();
-    std::chrono::milliseconds ms = std::chrono::duration_cast<std::chrono::milliseconds>(
-        response_received_time - downstream_request_complete_time_);
+    const uint64_t upstream_service_time_ms =
+        std::chrono::duration_cast<std::chrono::milliseconds>(response_received_time -
+                                                              downstream_request_complete_time_)
+            .count();
     if (!config_->suppress_envoy_headers_) {
-      headers->setEnvoyUpstreamServiceTime(ms.count());
+      headers->setEnvoyUpstreamServiceTime(upstream_service_time_ms);
     }
+    // Record the upstream service time in the stats
+    stats_.upstream_service_time_.recordValue(upstream_service_time_ms);
```

Comment:
one more question, how is this histogram different from `upstream_rq_time` on virtual cluster level?https://envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/router_filter#virtual-clusters

Reply:
I think this is the more important question. It would be better to avoid adding redundant histograms.